File: NetFramework/InputPanel/VB/TestCases/XmlUtil.vb
Imports System.Xml
Imports System.IO
Imports System.Runtime.Serialization.Formatters.Binary

''' <summary>
''' Utility class for reading Xml.
''' </summary>
Friend Class XmlUtil

    Shared Function Read(ByVal nd As XmlNode, ByVal attName As String, ByVal defVal As Boolean) As Boolean
        Dim att As XmlAttribute = nd.Attributes(attName)
        If att IsNot Nothing Then
            Return Boolean.Parse(att.Value)
        End If
        Return defVal
    End Function

    Shared Function Read(ByVal nd As XmlNode, ByVal attName As String, ByVal defVal As Int32) As Int32
        Dim att As XmlAttribute = nd.Attributes(attName)
        If att IsNot Nothing Then
            Return Int32.Parse(att.Value)
        End If
        Return defVal
    End Function

    Shared Function Read(ByVal nd As XmlNode, ByVal attName As String, ByVal defVal As String) As String
        Dim att As XmlAttribute = nd.Attributes(attName)
        If att IsNot Nothing Then
            Return att.Value
        End If
        Return defVal
    End Function

    Shared Function ReadImage(ByVal nd As XmlNode, ByVal prefix As String) As Image

        ' read image from file
        Dim att As XmlAttribute = nd.Attributes(prefix + "File")
        If att IsNot Nothing Then
            Dim fileName As String = att.Value
            If Not Path.IsPathRooted(fileName) And nd.BaseURI.Length > 0 Then
                Dim path As String = System.IO.Path.GetDirectoryName(nd.BaseURI)
                If path.StartsWith("file:\\") Then
                    path = path.Substring(6)
                End If
                fileName = System.IO.Path.Combine(path, fileName)
            End If
            If File.Exists(fileName) Then
                Return Image.FromFile(fileName)
            End If
        End If

        ' read image from base64 string
        att = nd.Attributes(prefix + "Base64")
        If att IsNot Nothing Then
            Dim buf() As Byte = Convert.FromBase64String(att.Value)
            Dim ms As MemoryStream = New MemoryStream(buf)
            Dim bf As BinaryFormatter = New BinaryFormatter()
            Return CType(bf.Deserialize(ms), Image)
        End If

        ' no image data in node
        Return Nothing
    End Function

End Class

File: NetFramework/InputPanel/CS/TestCases/XmlUtil.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Xml;
using System.Runtime.Serialization.Formatters.Binary;
using System.IO;

namespace TestCases
{
    /// <summary>
    /// Utility class for reading Xml.
    /// </summary>
    internal class XmlUtil
    {
        public static bool Read(XmlNode nd, string attName, bool defVal)
        {
            XmlAttribute att = nd.Attributes[attName];
            if (att != null)
            {
                return bool.Parse(att.Value);
            }
            return defVal;
        }
        public static int Read(XmlNode nd, string attName, int defVal)
        {
            XmlAttribute att = nd.Attributes[attName];
            if (att != null)
            {
                return int.Parse(att.Value);
            }
            return defVal;
        }
        public static string Read(XmlNode nd, string attName, string defVal)
        {
            XmlAttribute att = nd.Attributes[attName];
            if (att != null)
            {
                return att.Value;
            }
            return defVal;
        }
        public static Image ReadImage(XmlNode nd, string prefix)
        {
            // read image from file
            XmlAttribute att = nd.Attributes[prefix + "File"];
            if (att != null)
            {
                string fileName = att.Value;
                if (!Path.IsPathRooted(fileName) && nd.BaseURI.Length > 0)
                {
                    string path = Path.GetDirectoryName(nd.BaseURI);
                    if (path.StartsWith(@"file:\"))
                    {
                        path = path.Substring(6);
                    }
                    fileName = Path.Combine(path, fileName);
                }
                if (File.Exists(fileName))
                {
                    return Image.FromFile(fileName);
                }
            }

            // read image from base64 string
            att = nd.Attributes[prefix + "Base64"];
            if (att != null)
            {
                byte[] buf = Convert.FromBase64String(att.Value);
                MemoryStream ms = new MemoryStream(buf);
                BinaryFormatter bf = new BinaryFormatter();
                return bf.Deserialize(ms) as Image;
            }

            // no image data in node
            return null;
        }
    }
}
